statev2/state/src/applicator/peer_index.rs

```diff
@@ -0,0 +1,242 @@
+//! Applicator methods for the peer index, separated out for discoverability
+
+use std::str::FromStr;
+
+use crate::storage::db::DbTxn;
+
+use super::{
+    error::StateApplicatorError, Result, StateApplicator, CLUSTER_MEMBERSHIP_TABLE, PEER_INFO_TABLE,
+};
+use common::types::gossip::{PeerInfo, WrappedPeerId};
+use external_api::bus_message::{SystemBusMessage, NETWORK_TOPOLOGY_TOPIC};
+use itertools::Itertools;
+use libmdbx::RW;
+use state_proto::{AddPeers as AddPeersMsg, RemovePeer as RemovePeerMsg};
+
+impl StateApplicator {
+    // -------------
+    // | Interface |
+    // -------------
+
+    /// Add new peers to the peer index
+    pub fn add_peers(&self, msg: AddPeersMsg) -> Result<()> {
+        let tx = self
+            .db()
+            .new_write_tx()
+            .map_err(StateApplicatorError::Storage)?;
+
+        // Index each peer
+        for peer in msg.peers.into_iter() {
+            // Parse the peer info and mark a successful heartbeat
+            let peer_info = PeerInfo::try_from(peer).map_err(StateApplicatorError::Proto)?;
+            peer_info.successful_heartbeat();
```

Comment:
is it accurate to mark a successful heartbeat here? not sure where a raft node may receive the `AddPeersMsg` from, but I could see situations where it syncs a set of known peers from one such peer, this wouldn't really be the same as receiving a heartbeat from each of these peers

Reply:
Yeah this effectively resets the timeout so that the latency in the connection layer is not observed as a peer timeout. This gives the local peer time to heartbeat the discovered peer.